chapter.md

```diff
@@ -0,0 +1,966 @@
+  const propValue = <any[]> view(lensForProp, state);
+  return assocPath(
+    action.payload.path,
+    concat(propValue, [action.payload.value]),
+    state
+  );
+```
+
+Again,
+we're using Ramda functions to create a new array of skills
+given an old array and the thing we want to add:
+`lensForProp` is the location of the skills array in our state,
+`propValue` is its current value,
+and `concat` wrapped up in `assocPath` gives us the old skills
+with the new one appended.
+The additions to handle updating and removing skills have a similar shape.
+
+Extending our user interface is equally easy.
+We start with a helper function that creates and dispatches the required action:
+
+```ts
+addSkill() {
+  this.ngRedux.dispatch(addIntoArray({
+    value: undefined,
+    path: [ 'character', 'skills' ]
+  }));
+}
+```
+
+We have put this in the form component,
+but neither Angular nor good design principles strictly require this:
+we could have put it (and similar functions) in a file full of utilities.
+
+Now that we have the functions we need,
+we can write the HTML needed to put everything in front of the user:
+
+```html
+<label>Skills:</label>
+
+<div *ngFor="let skillSlot of characterForm.skills; let i = index;">
+  <select
+    [value]="skillSlot"
+    (change)="onSelectSkill($event, i)">
+    <option *ngFor="let skill of skills" [value]="skill">
+      {{ skill }}
+    </option>
+  </select>
+  <button type="button" (click)="removeSkill(i)">Remove</button>
+</div>
+
+<button type="button" (click)="addSkill()">Add skill</button>
+```
+
+The last line of this HTML is the one that lets users add skills;
+the rest is to handle skill display and removal.
+
+## The Payoff: Validation
+
+One way to see how these changes pays off is to look at validation,
+and in particular at the way in which we can use *selectors* to compute data by composing functions,
+and then use *memoization* to make state changes more efficient.
+
+Selectors chain functions together
+and pipe their return values into the last function in the chain.
+In the simple code below,
+the selector will return `form.character`:
+
+```ts
+const formStateSelector = (state: IAppState) => state.form;
+
+const characterFormSelector = createSelector(
+  formStateSelector,
+  (form: IForm) => form.character
+);
+```
+
+With that in hand,
+let's write a validation function that checks that all required fields are present
+and add it to our component:
+
+```ts
+export const isFormValid = createSelector(
+  characterFormSelector,
+  character => character.name
+    && character.bioSummary.age
+    && character.skills.length > 0
+);
+
+@select(isFormValidSelector)
+isFormValid$: Observable<boolean>;
+```
+
+Now,
+after selecting,
+we can use `isFormValid$` with an asynchronous pipe in our template:
+
+```ts
+<button 
+  [disabled]="!(isFormValid$ | async)"
+  type="submit">
+  Save
+</button>
+```
+
+(The `$` on the end of `isFormValid$` is a convention in the RxJS library meaning,
```

Comment:
It's a variable naming convention, hungarian notation style; often used with RxJS but not RxJS specific.